# Importing and Setting Up Python Modules

In [1]:
# The usual data visualization modules

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns
sns.set(style="darkgrid")

# This is to increase the size of all figures
plt.rcParams["figure.dpi"] = 150

In [2]:
# Modules to create interactive plots

import chart_studio.plotly as py
import cufflinks as cf

from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot

init_notebook_mode(connected=True)
cf.go_offline()

In [3]:
# Date & time manipulation modules

import datetime as dt
import datetime
from datetime import datetime

In [4]:
# Used to execute multple lines of code in a single cell

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [5]:
# # Used to display dataframes horizontally (instead of vertically)

# from IPython.display import display_html
# def display_side_by_side(*args):
#     html_str=''
#     for df in args:
#         html_str+=df.to_html()
#     display_html(html_str.replace('table','table style="display:inline"'),raw=True)

    
## ABOVE CODE IS NOT USED! Instead I used the solution here:
# https://stackoverflow.com/a/50899244

# Importing Raw Data Files

In [6]:
# This is the first set of data that was collected

ffx = pd.read_csv("ffxLighting_original_june28.txt")

In [7]:
# This is the (larger) second set of data

ffx2 = pd.read_csv("ffxLighting_new_july1.txt")

**Initial Look At The Data**

In [8]:
ffx.head()
ffx2.head()

count                      timings
0      1   2020-06-28 16:40:11.538763
1      2   2020-06-28 16:40:17.543537
2      3   2020-06-28 16:40:28.666688
3      4   2020-06-28 16:40:34.640885
4      5   2020-06-28 16:40:45.889237

count                      timings
0      1   2020-07-01 22:18:22.806651
1      2   2020-07-01 22:18:49.157322
2      3   2020-07-01 22:18:55.197844
3      4   2020-07-01 22:19:06.378970
4      5   2020-07-01 22:19:12.418819

**General Info About Raw Data**

In [9]:
ffx.describe()
ffx2.describe()

count
count  1135.000000
mean    568.000000
std     327.790584
min       1.000000
25%     284.500000
50%     568.000000
75%     851.500000
max    1135.000000

count
count  1859.000000
mean    930.000000
std     536.791393
min       1.000000
25%     465.500000
50%     930.000000
75%    1394.500000
max    1859.000000

In [10]:
ffx.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   count     1135 non-null   int64 
 1    timings  1135 non-null   object
dtypes: int64(1), object(1)
memory usage: 17.9+ KB


In [11]:
ffx2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1859 entries, 0 to 1858
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   count     1859 non-null   int64 
 1    timings  1859 non-null   object
dtypes: int64(1), object(1)
memory usage: 29.2+ KB


In [12]:
ffx.columns
ffx2.columns

Index(['count', ' timings'], dtype='object')

Index(['count', ' timings'], dtype='object')

**Looking At Specific Data Entries**

NOTES:
1. There is a leading space in the timings column (both in column header and the data values)!
2. The data in the 'timings' column are of type 'string'!

In [13]:
ffx["count"][0]
ffx[" timings"][0]

1

' 2020-06-28 16:40:11.538763'

In [14]:
type(ffx["count"][0])
type(ffx[" timings"][0])

numpy.int64

str

In [15]:
ffx2["count"][0]
ffx2[" timings"][0]

1

' 2020-07-01 22:18:22.806651'

In [16]:
type(ffx2["count"][0])
type(ffx2[" timings"][0])

numpy.int64

str

# **Data Cleaning**

### Part 1/2 - Original Data (ffx)

First, we shall get rid of the leading space in the 'timings' columns.

In [17]:
ffx.columns
ffx[" timings"][0]

Index(['count', ' timings'], dtype='object')

' 2020-06-28 16:40:11.538763'

In [18]:
# Eliminating leading space in the column header

ffx.rename(mapper={" timings": "timings"}, axis=1, inplace=True)
ffx.columns

Index(['count', 'timings'], dtype='object')

In [24]:
# Eliminating leading space in the data values

ffx["timings"] = ffx["timings"].apply(lambda string: string[1:])
ffx["timings"][0]
type(ffx["timings"][0])

'20-06-28 16:40:11.538763'

str

Next, we see if there are any null values. (I already know there aren't any, but we should still check to make sure.)

In [26]:
# Verifying that there are NO null values

ffx.count()
ffx.isnull().count()

count      1135
timings    1135
dtype: int64

count      1135
timings    1135
dtype: int64

Next, we convert the 'timings' data from string-type to Datetime-type.

In [29]:
# Converting 'timings' data to Datetime-type

ffx["Datetime"] = pd.to_datetime(ffx["timings"])

# Check to make sure it worked as expected
ffx.head()
ffx["Datetime"][0]
type(ffx["Datetime"][0])

count                   timings                   Datetime
0      1  20-06-28 16:40:11.538763 2028-06-20 16:40:11.538763
1      2  20-06-28 16:40:17.543537 2028-06-20 16:40:17.543537
2      3  20-06-28 16:40:28.666688 2028-06-20 16:40:28.666688
3      4  20-06-28 16:40:34.640885 2028-06-20 16:40:34.640885
4      5  20-06-28 16:40:45.889237 2028-06-20 16:40:45.889237

Timestamp('2028-06-20 16:40:11.538763')

pandas._libs.tslibs.timestamps.Timestamp

Next, I happen to know that some values at the *tail end* of the data are bad data (it was recorded while I was dismantling the data capture setup). We will find the bad data, and remove the respective rows.

NOTE: I also know that the data at the *front end* is okay, as I had correctly configured my data collection setup before I began to collect data.

In [31]:
# Initial check to see if we can find where the bad data starts

ffx[1100:]

count                   timings                   Datetime
1100   1101  20-06-28 19:17:45.300614 2028-06-20 19:17:45.300614
1101   1102  20-06-28 19:17:51.295674 2028-06-20 19:17:51.295674
1102   1103  20-06-28 19:18:02.442681 2028-06-20 19:18:02.442681
1103   1104  20-06-28 19:18:08.487168 2028-06-20 19:18:08.487168
1104   1105  20-06-28 19:18:19.587722 2028-06-20 19:18:19.587722
1105   1106  20-06-28 19:18:25.632636 2028-06-20 19:18:25.632636
1106   1107  20-06-28 19:18:36.770161 2028-06-20 19:18:36.770161
1107   1108  20-06-28 19:18:42.812585 2028-06-20 19:18:42.812585
1108   1109  20-06-28 19:18:53.944019 2028-06-20 19:18:53.944019
1109   1110  20-06-28 19:18:59.994721 2028-06-20 19:18:59.994721
1110   1111  20-06-28 19:19:11.182394 2028-06-20 19:19:11.182394
1111   1112  20-06-28 19:19:17.140761 2028-06-20 19:19:17.140761
1112   1113  20-06-28 19:19:28.370228 2028-06-20 19:19:28.370228
1113   1114  20-06-28 19:19:34.304837 2028-06-20 19:19:34.304837
1114   1115  20-06-28 19:19:45.509329 2028-06-20 19:19:45.509329
1115   1116  20-06-28 19:19:51.529480 2028-06-20 19:19:51.529480
1116   1117  20-06-28 19:20:02.690825 2028-06-20 19:20:02.690825
1117   1118  20-06-28 19:20:08.719560 2028-06-20 19:20:08.719560
1118   1119  20-06-28 19:20:18.647719 2028-06-20 19:20:18.647719
1119   1120  20-06-28 19:20:23.351210 2028-06-20 19:20:23.351210
1120   1121  20-06-28 19:20:40.833734 2028-06-20 19:20:40.833734
1121   1122  20-06-28 19:20:46.595614 2028-06-20 19:20:46.595614
1122   1123  20-06-28 19:20:48.694830 2028-06-20 19:20:48.694830
1123   1124  20-06-28 19:21:02.183183 2028-06-20 19:21:02.183183
1124   1125  20-06-28 19:21:09.733099 2028-06-20 19:21:09.733099
1125   1126  20-06-28 19:21:17.263125 2028-06-20 19:21:17.263125
1126   1127  20-06-28 19:21:30.427996 2028-06-20 19:21:30.427996
1127   1128  20-06-28 19:21:45.578473 2028-06-20 19:21:45.578473
1128   1129  20-06-28 19:21:45.679359 2028-06-20 19:21:45.679359
1129   1130  20-06-28 19:21:56.824975 2028-06-20 19:21:56.824975
1130   1131  20-06-28 19:23:27.663365 2028-06-20 19:23:27.663365
1131   1132  20-06-28 19:24:52.651587 2028-06-20 19:24:52.651587
1132   1133  20-06-28 19:24:58.391024 2028-06-20 19:24:58.391024
1133   1134  20-06-28 19:25:35.689479 2028-06-20 19:25:35.689479
1134   1135  20-06-28 19:25:41.456840 2028-06-20 19:25:41.456840

Based on our preliminary assessment, it's difficult to see which indices have the bad data. Since we are interested in investigating the time differences between lightning strikes, let's compute some quick preliminary time differences with the above data. This may give us some depth and allow us to find the bad data.

In [34]:
# Computing time differences
time_differences = [ (  f'{ffx["count"][i+1]}---{ffx["count"][i]}', 
                        ffx["Datetime"][i+1]-ffx["Datetime"][i]  ) for i in range(1100,1134) ]

# Converting above array into Pandas dataframe
bad_time_diff = pd.DataFrame(time_differences)

In [35]:
bad_time_diff

0               1
0   1102---1101 00:00:05.995060
1   1103---1102 00:00:11.147007
2   1104---1103 00:00:06.044487
3   1105---1104 00:00:11.100554
4   1106---1105 00:00:06.044914
5   1107---1106 00:00:11.137525
6   1108---1107 00:00:06.042424
7   1109---1108 00:00:11.131434
8   1110---1109 00:00:06.050702
9   1111---1110 00:00:11.187673
10  1112---1111 00:00:05.958367
11  1113---1112 00:00:11.229467
12  1114---1113 00:00:05.934609
13  1115---1114 00:00:11.204492
14  1116---1115 00:00:06.020151
15  1117---1116 00:00:11.161345
16  1118---1117 00:00:06.028735
17  1119---1118 00:00:09.928159
18  1120---1119 00:00:04.703491
19  1121---1120 00:00:17.482524
20  1122---1121 00:00:05.761880
21  1123---1122 00:00:02.099216
22  1124---1123 00:00:13.488353
23  1125---1124 00:00:07.549916
24  1126---1125 00:00:07.530026
25  1127---1126 00:00:13.164871
26  1128---1127 00:00:15.150477
27  1129---1128 00:00:00.100886
28  1130---1129 00:00:11.145616
29  1131---1130 00:01:30.838390
30  1132---1131 00:01:24.988222
31  1133---1132 00:00:05.739437
32  1134---1133 00:00:37.298455
33  1135---1134 00:00:05.767361

We can see a very clear alternating pattern in the time difference data above. We see that thsi pattern breaks starting with index 17 (of this particular array!). This means that the data points with a count of **1119 and onward** are the bad data. *Recall: This was when I was dismantling the data collection setup*

So, in our ffx dataframe we can delete the rows with 'count' starting at 1119 and greater. *This translates to dataframe indices of **1118** and greater!*

In [36]:
ffx.shape

(1135, 3)

In [40]:
# Dropping the bad data from 'ffx' dataframe

drop_condition = ffx["count"] >= 1119

ffx_cleaned = ffx.drop(ffx[drop_condition].index, axis=0)

In [41]:
ffx_cleaned

count                   timings                   Datetime
0         1  20-06-28 16:40:11.538763 2028-06-20 16:40:11.538763
1         2  20-06-28 16:40:17.543537 2028-06-20 16:40:17.543537
2         3  20-06-28 16:40:28.666688 2028-06-20 16:40:28.666688
3         4  20-06-28 16:40:34.640885 2028-06-20 16:40:34.640885
4         5  20-06-28 16:40:45.889237 2028-06-20 16:40:45.889237
...     ...                       ...                        ...
1113   1114  20-06-28 19:19:34.304837 2028-06-20 19:19:34.304837
1114   1115  20-06-28 19:19:45.509329 2028-06-20 19:19:45.509329
1115   1116  20-06-28 19:19:51.529480 2028-06-20 19:19:51.529480
1116   1117  20-06-28 19:20:02.690825 2028-06-20 19:20:02.690825
1117   1118  20-06-28 19:20:08.719560 2028-06-20 19:20:08.719560

[1118 rows x 3 columns]

### Part 2/2 - Second Dataset (ffx2)

We now essentially repeat the above process 'ffx2' dataframe.

In [43]:
ffx2.columns
ffx2[" timings"][0]

Index(['count', ' timings'], dtype='object')

' 2020-07-01 22:18:22.806651'

In [45]:
# Eliminating leading space in the column header

ffx2.rename(mapper={" timings": "timings"}, axis=1, inplace=True)
ffx2.columns

Index(['count', 'timings'], dtype='object')

In [46]:
# Eliminating leading space in the data values

ffx2["timings"] = ffx2["timings"].apply(lambda string: string[1:])
ffx2["timings"][0]
type(ffx2["timings"][0])

'2020-07-01 22:18:22.806651'

str

In [48]:
# Looking for null values

ffx2.count()
ffx2.isnull().count()

count      1859
timings    1859
dtype: int64

count      1859
timings    1859
dtype: int64

Next, we convert the 'timings' column to Datetime-type

In [49]:
# Converting to Datetime objects

ffx2["Datetime"] = pd.to_datetime(ffx2["timings"])

# Verifying it worked correctly
ffx2.head()
ffx2["Datetime"][0]
type(ffx2["Datetime"][0])

count                     timings                   Datetime
0      1  2020-07-01 22:18:22.806651 2020-07-01 22:18:22.806651
1      2  2020-07-01 22:18:49.157322 2020-07-01 22:18:49.157322
2      3  2020-07-01 22:18:55.197844 2020-07-01 22:18:55.197844
3      4  2020-07-01 22:19:06.378970 2020-07-01 22:19:06.378970
4      5  2020-07-01 22:19:12.418819 2020-07-01 22:19:12.418819

Timestamp('2020-07-01 22:18:22.806651')

pandas._libs.tslibs.timestamps.Timestamp

Now, for this second data set, I happen to know that there is bad data at the *front end* of this data set. This is due to my starting the data collection process, but then fiddling with the setup.

NOTE: I know the data at the *tail end* of this data set is good data, as my data collection program had crashed right when it collected the last data value.

In [55]:
# Initial look to see if we can eye-ball the bad data (at the front end)

ffx2[:20]

count                     timings                   Datetime
0       1  2020-07-01 22:18:22.806651 2020-07-01 22:18:22.806651
1       2  2020-07-01 22:18:49.157322 2020-07-01 22:18:49.157322
2       3  2020-07-01 22:18:55.197844 2020-07-01 22:18:55.197844
3       4  2020-07-01 22:19:06.378970 2020-07-01 22:19:06.378970
4       5  2020-07-01 22:19:12.418819 2020-07-01 22:19:12.418819
5       6  2020-07-01 22:19:23.520136 2020-07-01 22:19:23.520136
6       7  2020-07-01 22:19:29.611230 2020-07-01 22:19:29.611230
7       8  2020-07-01 22:19:40.801496 2020-07-01 22:19:40.801496
8       9  2020-07-01 22:19:46.788705 2020-07-01 22:19:46.788705
9      10  2020-07-01 22:19:57.940598 2020-07-01 22:19:57.940598
10     11  2020-07-01 22:20:03.941334 2020-07-01 22:20:03.941334
11     12  2020-07-01 22:20:15.103793 2020-07-01 22:20:15.103793
12     13  2020-07-01 22:20:21.131292 2020-07-01 22:20:21.131292
13     14  2020-07-01 22:20:32.341355 2020-07-01 22:20:32.341355
14     15  2020-07-01 22:20:38.373277 2020-07-01 22:20:38.373277
15     16  2020-07-01 22:20:49.596912 2020-07-01 22:20:49.596912
16     17  2020-07-01 22:20:55.627065 2020-07-01 22:20:55.627065
17     18  2020-07-01 22:21:06.831006 2020-07-01 22:21:06.831006
18     19  2020-07-01 22:21:12.843312 2020-07-01 22:21:12.843312
19     20  2020-07-01 22:21:24.005430 2020-07-01 22:21:24.005430

As expected, we do see that the first few data points are bad data as they do not follow the pattern. So, as previously done, we shall compute the time differences to see how many rows of data we need to remove.

In [57]:
# Computing time differences (at head of data)

time_differences_2 = [(f'{ffx2["count"][i+1]}---{ffx2["count"][i]}', ffx2["Datetime"][i+1]-ffx2["Datetime"][i]) for i in range(10)]

# Converting to dataframe for ease of use
bad_time_diff2_head = pd.DataFrame(time_differences_2)

In [58]:
bad_time_diff2_head

0               1
0    2---1 00:00:26.350671
1    3---2 00:00:06.040522
2    4---3 00:00:11.181126
3    5---4 00:00:06.039849
4    6---5 00:00:11.101317
5    7---6 00:00:06.091094
6    8---7 00:00:11.190266
7    9---8 00:00:05.987209
8   10---9 00:00:11.151893
9  11---10 00:00:06.000736

We see that the first time difference does not follow the observed pattern but the second time difference and beyond does. This means that only the first data point is the bad data. We will remove it.

In [59]:
# Removing (the single) bad data point at head of data set

ffx2_cleaned = ffx2.drop(labels=0, axis=0)
ffx2_cleaned

count                     timings                   Datetime
1         2  2020-07-01 22:18:49.157322 2020-07-01 22:18:49.157322
2         3  2020-07-01 22:18:55.197844 2020-07-01 22:18:55.197844
3         4  2020-07-01 22:19:06.378970 2020-07-01 22:19:06.378970
4         5  2020-07-01 22:19:12.418819 2020-07-01 22:19:12.418819
5         6  2020-07-01 22:19:23.520136 2020-07-01 22:19:23.520136
...     ...                         ...                        ...
1854   1855  2020-07-02 02:43:49.553087 2020-07-02 02:43:49.553087
1855   1856  2020-07-02 02:44:00.693588 2020-07-02 02:44:00.693588
1856   1857  2020-07-02 02:44:06.734364 2020-07-02 02:44:06.734364
1857   1858  2020-07-02 02:44:17.871546 2020-07-02 02:44:17.871546
1858   1859  2020-07-02 02:44:23.952632 2020-07-02 02:44:23.952632

[1858 rows x 3 columns]

**Question: Should we combine these two dataframes?**

**Answer:** First and most obviously, we know these two dataframes (ffx_cleaned & ffx2_cleaned) hold the exact same type of information, namely the count and timing data of lightning strikes. We also know that our main goal of this data analysis is to investigate the time differences in lightning strikes. 

This being said, the decision we will make is to **not combine the dataframes**. Here are my reasons:

1. It does not make sense to combine them because the data was collected on during different time spans.

2. If we were to combine the dataframes, the point where the two dataframes join, there will be a large discrepency in the time difference (infact, as large as a few days!).

Ultimately, the easier path is to analyze the dataframes separately. This will also allow us to **compare the results of two independent data sets**, which is always a good thing to do!

# Feature Engineering

### Part 1/2 - Original Dataset Cleaned (ffx_cleaned)

Of course, the main feature we wish to create is a "time difference" feature.

In [60]:
ffx_cleaned.head()

count                   timings                   Datetime
0      1  20-06-28 16:40:11.538763 2028-06-20 16:40:11.538763
1      2  20-06-28 16:40:17.543537 2028-06-20 16:40:17.543537
2      3  20-06-28 16:40:28.666688 2028-06-20 16:40:28.666688
3      4  20-06-28 16:40:34.640885 2028-06-20 16:40:34.640885
4      5  20-06-28 16:40:45.889237 2028-06-20 16:40:45.889237

In [73]:
# Creating the time difference feature

ffx_cleaned["deltaT"] = ffx_cleaned["Datetime"] - ffx_cleaned["Datetime"].shift(1)
ffx_cleaned.head()
type(ffx_cleaned["deltaT"][1])

count                   timings                   Datetime          deltaT
0      1  20-06-28 16:40:11.538763 2028-06-20 16:40:11.538763             NaT
1      2  20-06-28 16:40:17.543537 2028-06-20 16:40:17.543537 00:00:06.004774
2      3  20-06-28 16:40:28.666688 2028-06-20 16:40:28.666688 00:00:11.123151
3      4  20-06-28 16:40:34.640885 2028-06-20 16:40:34.640885 00:00:05.974197
4      5  20-06-28 16:40:45.889237 2028-06-20 16:40:45.889237 00:00:11.248352

pandas._libs.tslibs.timedeltas.Timedelta

Something that may be useful is to extract only the seconds information of each time difference. We create this feature below.

In [76]:
# Just for reference

ffx_cleaned["deltaT"][3]
ffx_cleaned["deltaT"][3].seconds
ffx_cleaned["deltaT"][3].microseconds

Timedelta('0 days 00:00:05.974197')

5

974197

In [77]:
# Creating the timedelta feature, in units of "seconds"

ffx_cleaned["dT(seconds)"] = [ffx_cleaned["deltaT"][i].seconds + ffx_cleaned["deltaT"][i].microseconds * 10**(-6) for i in range(0,len(ffx_cleaned["deltaT"]))]

In [78]:
ffx_cleaned

count                   timings                   Datetime  \
0         1  20-06-28 16:40:11.538763 2028-06-20 16:40:11.538763   
1         2  20-06-28 16:40:17.543537 2028-06-20 16:40:17.543537   
2         3  20-06-28 16:40:28.666688 2028-06-20 16:40:28.666688   
3         4  20-06-28 16:40:34.640885 2028-06-20 16:40:34.640885   
4         5  20-06-28 16:40:45.889237 2028-06-20 16:40:45.889237   
...     ...                       ...                        ...   
1113   1114  20-06-28 19:19:34.304837 2028-06-20 19:19:34.304837   
1114   1115  20-06-28 19:19:45.509329 2028-06-20 19:19:45.509329   
1115   1116  20-06-28 19:19:51.529480 2028-06-20 19:19:51.529480   
1116   1117  20-06-28 19:20:02.690825 2028-06-20 19:20:02.690825   
1117   1118  20-06-28 19:20:08.719560 2028-06-20 19:20:08.719560   

              deltaT  dT(seconds)  
0                NaT          NaN  
1    00:00:06.004774     6.004774  
2    00:00:11.123151    11.123151  
3    00:00:05.974197     5.974197  
4    00:00:11.248352    11.248352  
...              ...          ...  
1113 00:00:05.934609     5.934609  
1114 00:00:11.204492    11.204492  
1115 00:00:06.020151     6.020151  
1116 00:00:11.161345    11.161345  
1117 00:00:06.028735     6.028735  

[1118 rows x 5 columns]

### Part 2/2 - Second Dataset Cleaned (ffx2_cleaned)

We create the same two features with the second data set (ffx2_cleaned)

In [79]:
ffx2_cleaned.head()

count                     timings                   Datetime
1      2  2020-07-01 22:18:49.157322 2020-07-01 22:18:49.157322
2      3  2020-07-01 22:18:55.197844 2020-07-01 22:18:55.197844
3      4  2020-07-01 22:19:06.378970 2020-07-01 22:19:06.378970
4      5  2020-07-01 22:19:12.418819 2020-07-01 22:19:12.418819
5      6  2020-07-01 22:19:23.520136 2020-07-01 22:19:23.520136

In [80]:
# Creating the time difference features (as Timedelta objects)
ffx2_cleaned["deltaT"] = ffx2_cleaned["Datetime"] - ffx2_cleaned["Datetime"].shift(1)


#Creating the time difference features (in units of "seconds")
ffx2_cleaned["dT(seconds)"] = [ffx2_cleaned["deltaT"][i].seconds + ffx2_cleaned["deltaT"][i].microseconds * 10**(-6) for i in range(1,len(ffx2_cleaned["deltaT"])+1)]

In [81]:
ffx2_cleaned

count                     timings                   Datetime  \
1         2  2020-07-01 22:18:49.157322 2020-07-01 22:18:49.157322   
2         3  2020-07-01 22:18:55.197844 2020-07-01 22:18:55.197844   
3         4  2020-07-01 22:19:06.378970 2020-07-01 22:19:06.378970   
4         5  2020-07-01 22:19:12.418819 2020-07-01 22:19:12.418819   
5         6  2020-07-01 22:19:23.520136 2020-07-01 22:19:23.520136   
...     ...                         ...                        ...   
1854   1855  2020-07-02 02:43:49.553087 2020-07-02 02:43:49.553087   
1855   1856  2020-07-02 02:44:00.693588 2020-07-02 02:44:00.693588   
1856   1857  2020-07-02 02:44:06.734364 2020-07-02 02:44:06.734364   
1857   1858  2020-07-02 02:44:17.871546 2020-07-02 02:44:17.871546   
1858   1859  2020-07-02 02:44:23.952632 2020-07-02 02:44:23.952632   

              deltaT  dT(seconds)  
1                NaT          NaN  
2    00:00:06.040522     6.040522  
3    00:00:11.181126    11.181126  
4    00:00:06.039849     6.039849  
5    00:00:11.101317    11.101317  
...              ...          ...  
1854 00:00:06.050874     6.050874  
1855 00:00:11.140501    11.140501  
1856 00:00:06.040776     6.040776  
1857 00:00:11.137182    11.137182  
1858 00:00:06.081086     6.081086  

[1858 rows x 5 columns]

# **Exploratory Data Analysis**

Now that we have cleaned our datasets and created the main features of investigation, we move on to our EDA to see what patterns and information we can uncover.

Just as we had observed in the data cleaning section, we again immediately see that there seems to be a periodic pattern to the time difference feature. To start investigating this, let's extract some basic information about the time difference feature.

In [84]:
ffx_cleaned.head(10)

count                   timings                   Datetime          deltaT  \
0      1  20-06-28 16:40:11.538763 2028-06-20 16:40:11.538763             NaT   
1      2  20-06-28 16:40:17.543537 2028-06-20 16:40:17.543537 00:00:06.004774   
2      3  20-06-28 16:40:28.666688 2028-06-20 16:40:28.666688 00:00:11.123151   
3      4  20-06-28 16:40:34.640885 2028-06-20 16:40:34.640885 00:00:05.974197   
4      5  20-06-28 16:40:45.889237 2028-06-20 16:40:45.889237 00:00:11.248352   
5      6  20-06-28 16:40:51.855336 2028-06-20 16:40:51.855336 00:00:05.966099   
6      7  20-06-28 16:41:03.071391 2028-06-20 16:41:03.071391 00:00:11.216055   
7      8  20-06-28 16:41:09.055359 2028-06-20 16:41:09.055359 00:00:05.983968   
8      9  20-06-28 16:41:20.245578 2028-06-20 16:41:20.245578 00:00:11.190219   
9     10  20-06-28 16:41:26.305111 2028-06-20 16:41:26.305111 00:00:06.059533   

   dT(seconds)  
0          NaN  
1     6.004774  
2    11.123151  
3     5.974197  
4    11.248352  
5     5.966099  
6    11.216055  
7     5.983968  
8    11.190219  
9     6.059533

In [85]:
ffx2_cleaned.head(10)

count                     timings                   Datetime  \
1       2  2020-07-01 22:18:49.157322 2020-07-01 22:18:49.157322   
2       3  2020-07-01 22:18:55.197844 2020-07-01 22:18:55.197844   
3       4  2020-07-01 22:19:06.378970 2020-07-01 22:19:06.378970   
4       5  2020-07-01 22:19:12.418819 2020-07-01 22:19:12.418819   
5       6  2020-07-01 22:19:23.520136 2020-07-01 22:19:23.520136   
6       7  2020-07-01 22:19:29.611230 2020-07-01 22:19:29.611230   
7       8  2020-07-01 22:19:40.801496 2020-07-01 22:19:40.801496   
8       9  2020-07-01 22:19:46.788705 2020-07-01 22:19:46.788705   
9      10  2020-07-01 22:19:57.940598 2020-07-01 22:19:57.940598   
10     11  2020-07-01 22:20:03.941334 2020-07-01 22:20:03.941334   

            deltaT  dT(seconds)  
1              NaT          NaN  
2  00:00:06.040522     6.040522  
3  00:00:11.181126    11.181126  
4  00:00:06.039849     6.039849  
5  00:00:11.101317    11.101317  
6  00:00:06.091094     6.091094  
7  00:00:11.190266    11.190266  
8  00:00:05.987209     5.987209  
9  00:00:11.151893    11.151893  
10 00:00:06.000736     6.000736

In [88]:
ffx_cleaned.describe()

count                  deltaT  dT(seconds)
count  1118.000000                    1117  1117.000000
mean    559.500000  0 days 00:00:08.591925     8.591926
std     322.883106  0 days 00:00:02.587277     2.587278
min       1.000000  0 days 00:00:05.909511     5.909511
25%     280.250000  0 days 00:00:06.004127     6.004127
50%     559.500000  0 days 00:00:06.112921     6.112921
75%     838.750000  0 days 00:00:11.173472    11.173472
max    1118.000000  0 days 00:00:11.320335    11.320335

In [89]:
ffx2_cleaned.describe()

count                  deltaT  dT(seconds)
count  1858.000000                    1857  1857.000000
mean    930.500000  0 days 00:00:08.580934     8.580934
std     536.502718  0 days 00:00:02.589416     2.589416
min       2.000000  0 days 00:00:00.564302     0.564302
25%     466.250000  0 days 00:00:06.006229     6.006229
50%     930.500000  0 days 00:00:06.110464     6.110464
75%    1394.750000  0 days 00:00:11.169323    11.169323
max    1859.000000  0 days 00:00:11.352328    11.352328

So apparantly, there are two datapoints in the second dataset that are well below the lower mode?!?!?

In [ ]:
fig3, axes3 = plt.subplots(nrows=1, ncols=2, figsize=(15,3))
# axes3[0].set_ylim([0,0.00001])
# axes3[1].set_ylim([0,0.00001])

sns.rugplot(ffx_cleaned["dT(seconds)"], ax=axes3[0], height = 0.5)
sns.rugplot(ffx2_cleaned["dT(seconds)"], ax=axes3[1], height = 0.5)

# ffx_cleaned["dT(seconds)"].iplot(kind=ru)

We see there are some clear outliers from the modes in the second dataset. I guess we'll try removing these to clean the data more...

REASONING FOR REMOVING OUTLIERS (can I attribute them to a separate process?):
    -maybe the webcam messed up? (not likely...)
    -maybe there was fog at the same time as the lightning? (is likely, but would have seen more outliers then...)
    -maybe a ps4 notification popped up in the top left corner of the screen (update)? (can I check this somehow?)
    -looking at the first data set, we know that the "good data" should be clustered around ~6s and ~11s
    -hence, these outlier data points are truly bad data, so we are okay to discard them...?

In [ ]:
boom = ffx2_cleaned[((ffx2_cleaned["dT(seconds)"]>6.2)  & (ffx2_cleaned["dT(seconds)"]<11)) | (ffx2_cleaned["dT(seconds)"] < 4)]
boom
# boom.sort_values("count", axis=1)

In [ ]:
ffx2_cleaned.shape
ffx2_cleaned.drop(labels=[902,979,903,978,901], axis=0, inplace=True)
ffx2_cleaned.shape

In [ ]:
ffx2_cleaned

In [ ]:
fig3, axes3 = plt.subplots(nrows=1, ncols=2, figsize=(15,3))

sns.rugplot(ffx_cleaned["dT(seconds)"], ax=axes3[0], height = 0.5)
sns.rugplot(ffx2_cleaned["dT(seconds)"], ax=axes3[1], height = 0.5)

Here's a really basic scatter plot

In [ ]:
ffx_cleaned.plot.scatter(x="count", y="dT(seconds)", c="red")
ffx2_cleaned.plot.scatter(x="count", y="dT(seconds)", c="blue")

We can isolate the higher and lower values in their own scatter plots:

In [ ]:
fig4, axes4 = plt.subplots(nrows=2, ncols=2, figsize=(12,12))

condition1S = ffx_cleaned['dT(seconds)']<8
condition1L = ffx_cleaned['dT(seconds)']>8
condition2S = ffx2_cleaned['dT(seconds)']<8
condition2L = ffx2_cleaned['dT(seconds)']>8

ffx_cleaned[condition1S].plot.scatter(x="count" , y="dT(seconds)" , c=[(1,0,0,0.3)], ax=axes4[0,0])
ffx_cleaned[condition1L].plot.scatter(x="count" , y="dT(seconds)" , c=[(1,0,0,0.9)], ax=axes4[0,1])
ffx2_cleaned[condition2S].plot.scatter(x="count" , y="dT(seconds)" , c=[(0,0,1,0.3)], ax=axes4[1,0])
ffx2_cleaned[condition2L].plot.scatter(x="count" , y="dT(seconds)" , c=[(0,0,1,0.9)], ax=axes4[1,1])


We'll start with looking at a basic statistical histogram

In [ ]:
fig, axes = plt.subplots(nrows=1, ncols=2)

sns.distplot(ffx_cleaned["dT(seconds)"], ax=axes[0])
sns.distplot(ffx2_cleaned["dT(seconds)"], ax=axes[1])

axes[0].set_title("first dataset")
axes[1].set_title("second dataset")

We can see some very obvious patterns here, namely that the data exhibits a bi-modal distribution! Let's isolate the data around each mode.

We start by extracting the modal data into a seperate dataframe, to make it easier to work with.

In [ ]:
data1 = ffx_cleaned[["count", "dT(seconds)"]]
data2 = ffx2_cleaned[["count", "dT(seconds)"]]

df1_styler = data1.head().style.set_table_attributes("style='display:inline'").set_caption('Dataset 1')
df2_styler = data2.head().style.set_table_attributes("style='display:inline'").set_caption('Dataset 2')

display_html(df1_styler._repr_html_()+" .....  "+df2_styler._repr_html_(), raw=True)

In [ ]:
data1_small_mode = data1[data1["dT(seconds)"]<8]
data1_large_mode = data1[data1["dT(seconds)"]>8]

data2_small_mode = data2[data2["dT(seconds)"]<8]
data2_large_mode = data2[data2["dT(seconds)"]>8]

data1_small_styler = data1_small_mode.head().style.set_table_attributes("style='display:inline'").set_caption('Data1 small')
data1_large_styler = data1_large_mode.head().style.set_table_attributes("style='display:inline'").set_caption('Data1 large')
data2_small_styler = data2_small_mode.head().style.set_table_attributes("style='display:inline'").set_caption('Data2 small')
data2_large_styler = data2_large_mode.head().style.set_table_attributes("style='display:inline'").set_caption('Data2 large')

display_html(data1_small_styler._repr_html_()+" "+data1_large_styler._repr_html_()+" "+data2_small_styler._repr_html_()+" "+data2_large_styler._repr_html_(), raw=True)

We can create box-and-whisker plots to see the quartiles???

Here are some histograms of the individual modes for each data set

In [ ]:
fig2, axes2 = plt.subplots(nrows=2, ncols=2, figsize=(12,12))
# plt.tight_layout()
# axes2

sns.distplot(data1_small_mode["dT(seconds)"], ax=axes2[0,0])
sns.distplot(data1_large_mode["dT(seconds)"], ax=axes2[0,1])
sns.distplot(data2_small_mode["dT(seconds)"], ax=axes2[1,0])
sns.distplot(data2_large_mode["dT(seconds)"], ax=axes2[1,1])

# Need to specify domain for data2 for some reason...
axes2[1,0].set_xlim([5.85,6.15])


axes2[0,0].set_title("first dataset (small mode)")
axes2[0,1].set_title("first dataset (large mode)")
axes2[1,0].set_title("second dataset (small mode)")
axes2[1,1].set_title("second dataset (large mode)")

We can see a more normal distribution! So obviously, let's get the mean and standard deviation!

In [ ]:
mean1_small = np.mean(data1_small_mode["dT(seconds)"])
std1_small = np.std(data1_small_mode["dT(seconds)"])

mean1_large = np.mean(data1_large_mode["dT(seconds)"])
std1_large = np.std(data1_large_mode["dT(seconds)"])

mean2_small = np.mean(data2_small_mode["dT(seconds)"])
std2_small = np.std(data2_small_mode["dT(seconds)"])

mean2_large = np.mean(data2_large_mode["dT(seconds)"])
std2_large = np.std(data2_large_mode["dT(seconds)"])

stats = {'small1':(mean1_small,std1_small),
         'large1':(mean1_large,std1_large),
         'small2':(mean2_small,std2_small),
         'large2':(mean2_large,std2_large)}

print("(mean, std):")
stats

We can see that the standard deviations are VERY SMALL, meaning that the time differences are very closely centered about the mean.

We can compute the coefficient of variance: CoV = std/mean

In [ ]:
cov1_small = stats['small1'][1]/stats['small1'][0]
cov1_large = stats['large1'][1]/stats['large1'][0]
cov2_small = stats['small2'][1]/stats['small2'][0]
cov2_large = stats['large2'][1]/stats['large2'][0]

cov1_small
cov1_large
cov2_small
cov2_large

Indeed, all our distributions have extremely low variance in general.

**Next, we see there may be a pattern in the data - the time differences seem to be alternating. Let's investigate and prove!**

In [ ]:
ffx_cleaned.head()
ffx2_cleaned.head()

We determine if data is alternating by checking if all corresponding count's are either even or odd.

In [ ]:
# This function is no good anymore!!!
# It won't work for the newer data, because I removed the outliers
def all_even_or_odd(array):

    remainders = list(map(lambda x: x % 2, array))
    
    if remainders.count(1) == len(array):
        return "The array contains all odd integers"
    elif remainders.count(0) == len(array):
        return "The array contains all even integers"
    else:
        return "The array contains both even and odd integers"

In [ ]:
# Function to determine integer parity

def alternating(array):
    """To check whether all elements are even or odd"""
    
    num = 0
    for i in range(len(array)):
        if (array[i]>8):
            num+=1
        else:
            num-=1
        
        if (num == 2 or num == -2):
            return "There is some point where the array is not alternating"
    
    return "The array is alternating!"

In [ ]:
ffx_cleaned["dT(seconds)"]
alternating(ffx_cleaned["dT(seconds)"].values[1:])

In [ ]:
ffx2_cleaned["dT(seconds)"][900:1000]
alternating(ffx2_cleaned["dT(seconds)"].values[1:])

In [ ]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):  # more options can be specified also
    ffx2_cleaned["dT(seconds)"][890:1000]

So, we see that counts 978 & 979 were removed (they were outliers), which causes the time differences to go from 5.987260 to 6.048158, without an alternating value in between. The values that were taken out where 8.622982 & 2.472409, respectively. 

It is reasonable to assume that if these outliers did not exist, this sequence of numbers would still be alternating.